### DISCUSSION TITLE
アプリケーション設計に関する理解を深めたい。

### DISCUSSION THREAD

**Discussion by s1-haya** [42-pong/42-pong#8] [Q&A]
アプリケーション設計に関する理解を深めたいのですが、おすすめの記事があれば教えてください！

**Comment by ayase-mstk** [comment]
早速起票してくださりありがとうございます！

 「アプリケーション設計に関して」だとすこし広い範囲の質問になってしまうかと思います。
簡単にでいいと思うのでどこまで調べてどこら辺がわからないのかを教えていただけると、より良い返信ができるかと思います！（偉そうに聞こえてしまったら申し訳ないです。。）

アプリケーション設計というのが何を指すかにもよると思うので、とりあえず幾つかのキーワードと、良いリンクがあればそれを貼っておこうと思います。


開発全体の流れについては以下の記事が参考になるかもしれません。
- https://zenn.dev/nyanchu/articles/27a3f95d98df45

また開発の流れは大きく2つあるのかなと思っています。この2つは「先に全体を決めてしまうか、常に動くものを小さくても作りながら少しずつ機能追加していくかという2つの思想」くらいの理解です。（間違ってたらすみません）
- ウォーターフォール開発
- アジャイル開発


設計手法については、いくつかの観点があるかと思います。
- 設計手法自体（ここらへんは僕も言葉だけ知っていてあまりよくわかってません。）
  - テスト駆動開発
  - DDD（ドメイン駆動開発）
  - オブジェクト指向設計
  - ユースケース指向設計
- アーキテクチャの観点（ここはバックエンド側でどうするか決めるべきかと思います。）
  - 3層アーキテクチャ
    - https://qiita.com/os1ma/items/7a229585ebdd8b7d86c2
  - レイヤーアーキテクチャ
  - オニオンアーキテクチャ
  - ヘキサゴナルアーキテクチャ
  - クリーンアーキテクチャ
    - https://little-hands.hatenablog.com/entry/2017/10/04/231743
- ツール
  - UML

**Comment by s1-haya** [comment]
ご返信ありがとうございます。

投げやりな質問の仕方になってしまってすみません！chatgptみたいな感覚で質問しちゃいました。これから、簡易的に調べたことをまとめて、わからないことがあれば具体的に質問しますね！

**Comment by s1-haya** [comment]
### アプリケーション設計とは
アプリケーションのアーキテクチャや構造を決めることに焦点当てたソフトウェア設計（Software Design）の一種

### ソフトウェア設計とは
ソフトウェアシステムがどのように機能するかを構想するプロセスのこと。このプロセスでは、実現したいソフトウェアに対して要件定義を行い、それに基づいて基本設計（high-level software architecture）と詳細設計（low-level component and algorithm design）を行う。ソフトウェア設計は、実装や変更の前に行われ、システムの目標を達成するための問題解決や計画を含む。

> Software design is the process of conceptualizing how a [software system](https://en.wikipedia.org/wiki/Software_system) will work before it is [implemented](https://en.wikipedia.org/wiki/Implementation) or modified

> Software design usually is directed by [goals](https://en.wikipedia.org/wiki/Goal) for the resulting system and involves problem-solving and planning – including both high-level [software architecture](https://en.wikipedia.org/wiki/Software_architecture) and low-level component and [algorithm design](https://en.wikipedia.org/wiki/Algorithm_design).

https://en.wikipedia.org/wiki/Software_design

### ソフトウェアとは
ソフトとウェアの造語。ソフトは変更が容易。ウェアは製品。ソフトウェアが発明された理由はマシンの振る舞いを素早く簡単に変更する必要があったから。

クリーンアーキテクチャの本の著者曰く、ソフトウェアシステムには２種類の価値がある。振る舞いの価値と構造の価値。
- 振る舞いの価値: システムの機能や動作に関する価値
- 構造の価値: システムが将来にわたって柔軟に変更可能で、保守や拡張が容易であるための設計に関連する価値

振る舞いの柔軟性はシステムの形状、コンポーネントの配置、コンポーネントの相互接続によって大きく依存するため、アーキテクチャが必要。著者はアイゼンハワーのマトリックスを使って振る舞いとアーキテクチャの関係性について次のように表した。

- 振る舞いは緊急だが、常に重要とは限らない。
- アーキテクチャは重要だが、常に緊急とは限らない。

| 緊急度 / 重要度 |  |  |
| -- | -- | -- |
|  | **重要かつ緊急** | **重要だが緊急でない** |
|  | **緊急だが重要でない** | **重要でも緊急でもない** |

重要かつ緊急なことと緊急だが重要でないことの区別し、機能の緊急性よりもアーキテクチャの重要視すべき。って書かれてた。

### アーキテクチャとは
ソフトウェアシステムを理解するために必要な構造の集合と、それらの構造やシステムを作成するための手法のこと。各構造は、ソフトウェアの要素、要素間の関係、さらに要素と関係の特性から成り立つ。

> Software architecture is the set of structures needed to reason about a [software system](https://en.wikipedia.org/wiki/Software_system) and the discipline of creating such structures and systems. Each structure comprises software elements, relations among them, and properties of both elements and relations

https://en.wikipedia.org/wiki/Software_architecture

**Comment by s1-haya** [comment]
最初、application designって検索すると、検索結果があんまりなかったので、早瀬さんから共有していただいた記事を元に調べた。認識ずれてたら教えてください。早瀬さんのアプリケーション設計っていう言葉に対する理解度深すぎる。

**Reply by ayase-mstk** [reply]
いえいえ、僕もこの前ちょっと調べただけで実は一つ一つに対してはそんなに深くわかってないです笑

[system-design-primer](https://github.com/donnemartin/system-design-primer/tree/master)
こちらのリンクは今回の設計で使うような知識があるかはわかりませんが、一般的に企業面接対策とかに必要な知識について書いてあったり、システム全体像の図もあったりしていいかもしれません。
